core/src/main/java/org/apache/druid/math/expr/BinaryLogicalOperatorExpr.java

```diff
@@ -57,6 +57,17 @@ protected final double evalDouble(double left, double right)
     // Use Double.compare for more consistent NaN handling.
     return Evals.asDouble(Double.compare(left, right) < 0);
   }
+
+  @Nullable
+  @Override
+  public ExprType getOutputType(InputBindingTypes inputTypes)
+  {
+    ExprType implicitCast = super.getOutputType(inputTypes);
+    if (ExprType.STRING.equals(implicitCast)) {
+      return ExprType.LONG;
+    }
+    return implicitCast;
```

Comment:
Looking at how this function works got me thinking about some stuff... Does this function need to be in sync with the behavior in `BinaryEvalOpExprBase#eval` (I think so 🤔) Since the `eval` method isn't implemented here, would it be better to implement it in `BinaryEvalOpExprBase`?

Can you explain how `getOutputType` would deal with default null handling mode.

Also, what does it mean to have an output type of null?

Reply:
>Looking at how this function works got me thinking about some stuff... Does this function need to be in sync with the behavior in BinaryEvalOpExprBase#eval (I think so 🤔) Since the eval method isn't implemented here, would it be better to implement it in BinaryEvalOpExprBase?

Yeah, the behavior of `getOutputType` should always match the behavior of `eval`. In this case, it can't really be pushed down because the math operators also implement `BinaryEvalOpExprBase`, but do not handle string inputs as numerical outputs. We could put another type in between these logical operators and `BinaryEvalOpExprBase` that provides it though, I will consider doing that.

>Can you explain how getOutputType would deal with default null handling mode.

`getOutputType` should be fully independent of how `druid.generic.useDefaultValueForNull` is set since it does not capture (currently) whether or not nulls can happen.

>Also, what does it mean to have an output type of null?

An output type of null signifies that we couldn't compute what the output type is, most likely because input type information isn't available. When the input to the expressions are actual segments (`QueryableIndexStorageAdapter`) then type information should always be available, and a `null` signifies an input that doesn't exist, but other adapters (and other usages of `Expr` such as for transforms) might not always have complete type information.

I will add all of this stuff to the javadocs.

Reply:
what should be the expected output type here if the left output type is a string and the right output type is a double? 
As per `org.apache.druid.math.expr.BinaryEvalOpExprBase#eval` it seems, the runtime output type will be `double` . As per `ExprType.autoTypeConversion` it seems to be `string`.

Reply:
Good eye, you spotted a thing I was purposefully ignoring because I think it is not really great behavior, or consistent with other math functions. The default behavior for those 2 input math operators is to consider the arguments as doubles as you've noticed, regardless of whether or not they are both doubles. I suspect it is implemented like this to allow null values inputs to still work as zeros in 'default' mode (`druid.generic.useDefaultValueForNull=true`), because the current expressions are not strongly typed so these nulls all end up as string values. See [this comment](https://github.com/apache/druid/pull/10370#discussion_r486762521) for a bit more explanation, and [this comment](https://github.com/apache/druid/pull/10084#discussion_r446486836) for some of the problems that this causes.

It is also a bit unintuitive behavior. A string column which contains numeric values will work in one of these math operators, so something like `2.0 + '3.1' = 5.1` works magically. But it will also potentially treat the string value as 0 if the string is not a number in default mode, so `2.0 + 'foo' = 2.0` (or null in sql compatible mode). The math functions (`min`, `max`, etc) instead treat either input as string as a 0/null output without evaluating the function.

I'm not actually sure what the best behavior here is, the math function behavior seems a bit more intuitive to me, where either input being a string produces a null output, so I chose to use that here. This PR is setting up for vectorized expressions, which are going to be strongly typed before processing, which I think makes this inconsistent/confusing/magical behavior not necessary. I was planning to raise a discussion about this part in a future PR since it isn't actually affecting anything yet, but I think it is good to start talking about it even though it isn't wired up to anything yet.

Also, I really think 'default' mode complicates the expression system quite a lot. It would be my preference for the expression system to always behave in an SQL compatible manner, and default mode only come into play on the boundaries for data coming in and going out, but I haven't fully thought through the implications of this and it requires some discussion I think, and might be a bit dramatic of a change.

Reply:
>so something like 2.0 + '3.1' = 5.1

That might not have been a great example, this might normal behavior in some other databases 😅. I'm going to have a closer read over https://www.postgresql.org/docs/9.0/typeconv-oper.html and https://www.postgresql.org/docs/9.0/typeconv-func.html and maybe reference some other docs too and see if might make sense to different versions of this implicit conversion function for different contexts.

Reply:
can you elaborate more on this? How will that look like? 
> This PR is setting up for vectorized expressions, which are going to be strongly typed before processing, which I think makes this inconsistent/confusing/magical behavior not necessary.
While I understand the right behavior is debatable, it may still be best to keep the type in sync with eval even if it's not very intuitive. It may require that the logic of figuring out the type is not reusable as a function `ExprType.autoTypeConversion` and may have to be written differently for some operators.  We can, later on, change the different functions, fixing both the `eval` and `getOutputType` together. But till then, they will be in sync.

Reply:
>can you elaborate more on this? How will that look like?

Ah, its going to be based on using these methods, just the processors will be specialized to deal with the correct type based on the set of input types. Since the output type information isn't used for non-vectorized expressions, I'm trying to model this as the change I want to see in the world and will ensure that this matches the behavior of the vectorized expressions, but I've gone ahead and split operator from function auto conversion and changed it to match existing behavior for now in case it is necessary, and can always consolidate them again in the future.

Reply:
I'm going to merge this PR, but we can continue this discussion in the next one, so this stuff can actually be tied into things in a concrete manner. Looking at postgres, its math functions do, at least somewhat, allow for implicit conversion of sting numbers to numbers. So `2.0 + '3.1' = 5.1` and `cos('1')` work, but `'2.0' + '3.1'` does not. In druid  currently the first would have the same behavior as postgres, but the 2nd would be null or 0 depending on the value of `druid.generic.useDefaultValueForNull`, while the 3rd expression would do string concatenation.

Reply:
Thanks @clintropolis